src/RequestParser.cpp

```diff
@@ -0,0 +1,724 @@
+ * clears the URI fragment, path, and query, sets the version to an
+ * empty string, clears the headers and body, sets the error code to 0,
+ * and indicates that the connection should not be closed.
+ *
+ * @param request The HTTPRequest object to be cleared.
+ */
+void RequestParser::clearRequest(HTTPRequest& request)
+{
+	request.method = MethodCount;
+	request.uri.fragment = "";
+	request.uri.path = "";
+	request.uri.query = "";
+	request.version = "";
+	request.headers.clear();
+	request.body = "";
+	request.errorCode = StatusOK;
+	request.shallCloseConnection = false;
+}
+
+/**
+ * @brief Clears the internal state of the RequestParser object.
+ *
+ * This function resets the internal state of the RequestParser by
+ * setting the flags `m_hasBody` and `m_chunked` to `false`, and
+ * clearing the contents of the `m_requestStream` stringstream.
+ */
+void RequestParser::clearParser()
+{
+	m_hasBody = false;
+	m_chunked = false;
+	m_requestStream.clear();
+	m_requestStream.str("");
+}
+
+/* ====== CONSTRUCTOR/DESTRUCTOR ====== */
+
+/**
+ * @brief Constructs a new RequestParser object.
+ *
+ * This constructor initializes the RequestParser object with default values.
+ * The `m_hasBody` member is set to `false` and the `m_chunked` member is set to `false`.
+ */
+RequestParser::RequestParser()
+	: m_hasBody(false)
+	, m_chunked(false)
+{
+}
+
+/* ====== MEMBER FUNCTIONS ====== */
+
+/**
+ * @brief Parses an HTTP request string into an HTTPRequest object.
+ *
+ * This function takes a raw HTTP request string and parses it into an
+ * `HTTPRequest` object. The parsing process involves several steps:
+ * 1. Parsing the request line (method, URI, and version).
+ * 2. Parsing the headers.
+ * 3. Parsing the body (if any).
+ *
+ * If any errors are encountered during parsing, an appropriate error code is
+ * set, and a `std::runtime_error` is thrown.
+ *
+ * @param requestString The raw HTTP request string to be parsed.
+ * @param request The HTTP request object to be filled.
+ * @throws std::runtime_error If there is any error during parsing.
+ */
+void RequestParser::parseHttpRequest(const std::string& requestString, HTTPRequest& request)
+{
+	m_requestStream.str(requestString);
+
+	// Step 1: Parse the request-line
+	std::string requestLine;
+	if (!std::getline(m_requestStream, requestLine) || requestLine.empty()) {
+		request.errorCode = StatusBadRequest;
+		throw std::runtime_error(ERR_MISS_REQUEST_LINE);
+	}
+	requestLine = parseMethod(requestLine, request);
+	requestLine = checkForSpace(requestLine, request);
+	requestLine = parseUri(requestLine, request);
+	requestLine = checkForSpace(requestLine, request);
+	requestLine = parseVersion(requestLine, request);
+	checkForCRLF(requestLine, request);
+
+	// Step 2: Parse headers
+	std::string headerLine;
+	// The end of the headers section is marked by an empty line (\r\n\r\n).
+	while (std::getline(m_requestStream, headerLine) && headerLine != "\r" && !headerLine.empty()) {
```

Comment:
!(std::getline(m_requestStream, headerLine)).fail() to explicitly convert to bool

Reply:
Changed.